sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseReadSplittableDoFn.java

```diff
@@ -0,0 +1,115 @@
+import org.apache.hadoop.hbase.TableName;
+import org.apache.hadoop.hbase.client.Connection;
+import org.apache.hadoop.hbase.client.ConnectionFactory;
+import org.apache.hadoop.hbase.client.Result;
+import org.apache.hadoop.hbase.client.ResultScanner;
+import org.apache.hadoop.hbase.client.Scan;
+import org.apache.hadoop.hbase.client.Table;
+
+/** A SplittableDoFn to read from HBase. */
+@BoundedPerElement
+class HBaseReadSplittableDoFn extends DoFn<Void, Result> {
+  private final SerializableConfiguration serializableConfiguration;
+  private final String tableId;
+  private final SerializableScan scan;
+
+  private Connection connection;
+
+  HBaseReadSplittableDoFn(
+      SerializableConfiguration serializableConfiguration, String tableId, SerializableScan scan) {
+    this.serializableConfiguration = serializableConfiguration;
+    this.tableId = tableId;
+    this.scan = scan;
+  }
+
+  @Setup
+  public void setup() throws Exception {
+    if (connection == null) {
+      connection = ConnectionFactory.createConnection(serializableConfiguration.get());
+    }
+  }
+
+  @ProcessElement
+  public void processElement(ProcessContext c, ByteKeyRangeTracker tracker) throws Exception {
+    TableName tableName = TableName.valueOf(tableId);
+    Table table = connection.getTable(tableName);
+    final ByteKeyRange range = tracker.currentRestriction();
+    Scan newScan =
+        new Scan(scan.get())
+            .setStartRow(range.getStartKey().getBytes())
+            .setStopRow(range.getEndKey().getBytes());
+    try (ResultScanner scanner = table.getScanner(newScan)) {
+      for (Result result : scanner) {
+        ByteKey key = ByteKey.copyFrom(result.getRow());
+        if (!tracker.tryClaim(key)) {
+          return;
+        }
+        c.output(result);
+      }
+      tracker.markDone();
+    }
+  }
+
+  @GetInitialRestriction
+  public ByteKeyRange getInitialRestriction(Void v) {
+    return ByteKeyRange.of(
+        ByteKey.copyFrom(scan.get().getStartRow()), ByteKey.copyFrom(scan.get().getStopRow()));
+  }
+
+  @SplitRestriction
+  public void splitRestriction(Void v, ByteKeyRange range, OutputReceiver<ByteKeyRange> receiver)
+      throws Exception {
+    // TODO remove once BEAM-4016 is fixed (lifecycle ignores calling setup before splitRestriction)
+    setup();
+
+    List<ByteKeyRange> splitRanges =
+        HBaseUtils.getRanges(
+            HBaseUtils.getRegionLocations(connection, tableId, scan.get()), tableId, scan.get());
+    if (splitRanges.size() > 1) {
+      for (ByteKeyRange splitRange : splitRanges) {
+        receiver.output(ByteKeyRange.of(splitRange.getStartKey(), splitRange.getEndKey()));
+      }
+      return;
+    }
+    receiver.output(range);
+  }
+
+  @NewTracker
+  public ByteKeyRangeTracker newTracker(ByteKeyRange range) {
+    return ByteKeyRangeTracker.of(range);
+  }
+
+  @Teardown
+  public void tearDown() throws Exception {
+    if (connection != null) {
```

Comment:
I think it can't be null here either

Reply:
It can't. but if it happens and I call `connection.close()` and it is null it will throw `NullPointerExcepton`. This is a pretty common defensive pattern in most of the IOs who do Teardown (at least in the WriteDoFns).